File: RateLimiter2/SlidingWindowCounter.js
const RateLimiterStrategy = require('./RateLimiterStrategy');

class SlidingWindowCounter extends RateLimiterStrategy {
  constructor(limit, windowMs) {
    super();
    this.limit = limit;
    this.windowMs = windowMs;
    this.requests = new Map();
  }

  isAllowed(ip) {
    const currentTime = Date.now();

    if (!this.requests.has(ip)) {
      this.requests.set(ip, {
        previousStart: currentTime - this.windowMs,
        previousCount: 0,
        currentStart: currentTime,
        currentCount: 0,
      });
    }

    const window = this.requests.get(ip);

    // Let's assume current window started 4 seconds ago, so timeLapsed = 4
    let timeLapsed = currentTime - window.currentStart;

    // If timeLapsed >= windowMs, time to reset the window
    if (timeLapsed >= this.windowMs) {
      window.previousStart = window.currentStart;
      window.previousCount =
        timeLapsed >= this.windowMs * 2 ? 0 : window.currentCount;
      window.currentStart = currentTime;
      window.currentCount = 0;

      this.requests.set(ip, window);

      // Fresh window started
      timeLapsed = 0;
    }

    // 3 / 8 = ~0.4 (~40% time lapsed) ---- Ratio: 0-0.9
    const ratio = timeLapsed / this.windowMs;

    // 1 - 0.4 = 0.6 (skip ~40% requests, take ~60% of previousCount to take 1 entire window i.e. ~60% of previous and ~40% of current)
    const effectivePreviousCount = window.previousCount * (1 - ratio);
    const effectiveTotal = window.currentCount + effectivePreviousCount;
    if (effectiveTotal < this.limit) {
      window.currentCount++;
      this.requests.set(ip, window);
      return true;
    }

    return false;
  }
}

module.exports = SlidingWindowCounter;

File: Go/RateLimiter2-go/sliding_window_counter.go
package main

import (
	"sync"
	"time"
)

type slidingWindowState struct {
	PreviousStart int64
	PreviousCount int
	CurrentStart  int64
	CurrentCount  int
}

type SlidingWindowCounter struct {
	mu       sync.Mutex
	Limit    int
	WindowMs int64
	Requests map[string]slidingWindowState
}

func NewSlidingWindowCounter(limit int, windowMs int64) *SlidingWindowCounter {
	return &SlidingWindowCounter{
		Limit:    limit,
		WindowMs: windowMs,
		Requests: make(map[string]slidingWindowState),
	}
}

func (s *SlidingWindowCounter) IsAllowed(ip string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	currentTime := time.Now().UnixMilli()

	window, ok := s.Requests[ip]
	if !ok {
		s.Requests[ip] = slidingWindowState{
			PreviousStart: currentTime - s.WindowMs,
			CurrentStart:  currentTime,
		}
		window = s.Requests[ip]
	}

	timeLapsed := currentTime - window.CurrentStart
	if timeLapsed >= s.WindowMs {
		window.PreviousStart = window.CurrentStart
		if timeLapsed >= s.WindowMs*2 {
			window.PreviousCount = 0
		} else {
			window.PreviousCount = window.CurrentCount
		}
		window.CurrentStart = currentTime
		window.CurrentCount = 0
		s.Requests[ip] = window
		timeLapsed = 0
	}

	ratio := float64(timeLapsed) / float64(s.WindowMs)
	effectivePreviousCount := float64(window.PreviousCount) * (1 - ratio)
	effectiveTotal := float64(window.CurrentCount) + effectivePreviousCount
	if effectiveTotal < float64(s.Limit) {
		window.CurrentCount++
		s.Requests[ip] = window
		return true
	}
	return false
}
